### PR TITLE
Use this only if the scope is null (not falsy)

### LINKED ISSUE #1958: 0.times.size returns null in strict mode
Opal generates the following code for `times`:

```js
return $send(self, 'enum_for', ['times'], ($$61 = function () {
  var self = $$61.$$s || this

  if ($$61.$$is_lambda || $$61.$$define_meth) {
    var $arity = arguments.length
    if ($arity !== 0) { Opal.block_ac($arity, 0, 'times') }
  }
  return self
}, $$61.$$s = self, $$61.$$arity = 0, $$61.$$parameters = [], $$61))
```

The problem is that `$$61.$$s` is falsy (the value is 0) so we are using `this` instead. But `this` in strict mode is `null`.

Thus, the following returns null instead of 0:

```js
(0).$times().$size() // returns null
```

### PR DESCRIPTION
**Pull request by Mogztter**
Resolves #1958

### CODE REVIEW

File: tasks/testing.rake [hunk 1]
```diff
@@ -335,7 +336,7 @@ platforms.each do |platform|
         includes = "-Itest -Ilib -Ivendored-minitest #{includes}"
 
         sh "ruby -rbundler/setup "\
-         "exe/opal #{includes} #{stubs} -R#{platform} -Dwarning -A --enable-source-location #{filename}"
+         "exe/opal #{includes} #{stubs} -R#{platform} -Dwarning -A --enable-strict-mode --enable-source-location #{filename}"
```
Comment (Mogztter) [inline tasks/testing.rake]: Enable strict mode in minitest

File: lib/opal/nodes/iter.rb [hunk 2]
```diff
@@ -17,7 +17,7 @@ def compile
 
         in_scope do
           identity = scope.identify!
-          add_temp "self = #{identity}.$$s || this"
+          add_temp "self = #{identity}.$$s == null ? this : #{identity}.$$s"
```
Comment (Mogztter) [inline lib/opal/nodes/iter.rb]: The actual fix

File: test/opal/test_strict.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: false
+require "test/unit"
+
+class TestStrict < Test::Unit::TestCase
+  def test_strict_mode
+    assert_equal(0, 0.times.size)
```
Comment (Mogztter) [inline test/opal/test_strict.rb]: If you revert the oneline fix in `lib/opal/nodes/iter.rb`, `minitest` should fail.

File: lib/opal/cli_options.rb [hunk 4]
```diff
@@ -160,6 +160,10 @@ def initialize
         options[:enable_source_location] = true
       end
 
+      on('--enable-strict-mode', '--no-method-missing', 'Enable/Disable method missing') do
```
[the remark this answers is not in the document]
Comment (Mogztter) [reply lib/opal/cli_options.rb]: Fixed

File: test/opal/test_strict.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  end
+end
```
[the remark this answers is not in the document]
Comment (Mogztter) [reply test/opal/test_strict.rb]: Fixed

File: opal/corelib/strict.js [hunk 6]
```diff
@@ -0,0 +1 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (Mogztter) [reply opal/corelib/strict.js]: Fixed

File: opal/corelib/string/encoding.rb [hunk 7]
```diff
@@ -210,6 +210,10 @@ def force_encoding(encoding)
 
       if (encoding === self.encoding) { return self; }
 
+      if (typeof self === 'string') {
```
Comment (Mogztter) [inline opal/corelib/string/encoding.rb]: Same here but I'm not sure about this fix.
The error is `TypeError: Cannot create property 'encoding' on string ')'`.

**Comment by Mogztter** [comment]
@elia I've tried to do as little change as possible but I think we have to address the `String#force_encoding` first.
I'm pretty sure my hack is not accurate but at the same time encoding in Ruby and JavaScript does not work in the same way and I can't think of any other solution to this problem...

**Comment by elia** [comment]
Merged (except for the CLI option) in #2023

**Pull request closed without merging**